File: 지은/Lv.2/오픈채팅방.py
def solution(record):
    users_dict = {}
    order = []
    for rec in record:
        info = rec.split();
        state,uid = info[0], info[1]
        
        if state in ["Enter", "Change"]:
            username = info[2]
            users_dict[uid] = username
        if state != "Change":
            order.append((uid,state))
            
    res =[]
    for uid,state in order:
        if state == "Enter":
            res.append(f"{users_dict[uid]}님이 들어왔습니다.")
        elif state == "Leave":
            res.append(f"{users_dict[uid]}님이 나갔습니다.")
    
    return res


# 다른 사람의 풀이1
def solution(record):
    answer = []
    namespace = {}
    printer = {'Enter':'님이 들어왔습니다.', 'Leave':'님이 나갔습니다.'}
    for r in record:
        rr = r.split(' ')
        if rr[0] in ['Enter', 'Change']:
            namespace[rr[1]] = rr[2]

    for r in record:
        if r.split(' ')[0] != 'Change':
            answer.append(namespace[r.split(' ')[1]] + printer[r.split(' ')[0]])

    return answer

# 다른 사람의 풀이2
def solution(record):
    answer = []
    id_nic = {}
    for r in record:
        s = r.split()
        if len(s) > 2:
            id_nic[s[1]] = s[2]

    for r in record:
        s = r.split()
        if s[0] == "Enter":
            answer.append(id_nic[s[1]] + "님이 들어왔습니다.")
        elif s[0] == "Leave":
            answer.append(id_nic[s[1]] + "님이 나갔습니다.")
    return answer

File: 화연/Lv.2/오픈채팅방.js
/**
 * https://school.programmers.co.kr/learn/courses/30/lessons/42888
 * ㅋㅋㅋㅋㅋ 너무 복잡하게 풀었다! ㅠㅠ
 * @param {string[]} record
 * @returns {string[]}
 */
function solution(record) {
  const recordObj = {};
  const answer = [];
  let j = 0;
  for (let i = 0; i < record.length; i++) {
    const [msg, id, nickname] = record[i].split(" ");
    let nowRecord = recordObj[id] || {
      nickname,
      index: [],
    };
    if (msg === "Enter") {
      answer.push(`님이 들어왔습니다.`);
      nowRecord.index.push(j);
      nowRecord = {
        ...nowRecord,
        nickname,
      };
      j++;
    } else if (msg === "Leave") {
      answer.push(`님이 나갔습니다.`);
      nowRecord.index.push(j);
      j++;
    } else {
      nowRecord = {
        ...nowRecord,
        nickname,
      };
    }
    recordObj[id] = { ...nowRecord };
  }
  for (const key in recordObj) {
    const indexArr = recordObj[key].index;
    const nickname = recordObj[key].nickname;
    for (const index of indexArr) {
      answer[index] = nickname + answer[index];
    }
  }
  return answer;
}

/**
 * @param {string[]} record
 * @returns {string[]}
 */
function solution(record) {
  const userInfo = {};
  const action = [];
  const stateMapping = {
    Enter: "님이 들어왔습니다.",
    Leave: "님이 나갔습니다.",
  };

  record.forEach((v) => {
    const [state, id, nick] = v.split(" ");

    if (state !== "Change") {
      action.push([state, id]);
    }

    if (nick) {
      userInfo[id] = nick;
    }
  });

  return action.map(([state, uid]) => {
    return `${userInfo[uid]}${stateMapping[state]}`;
  });
}
